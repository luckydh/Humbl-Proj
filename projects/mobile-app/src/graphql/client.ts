import { ApolloClient, createHttpLink, InMemoryCache, from, ApolloLink } from "@apollo/client";
import { setContext } from "@apollo/client/link/context";
import { onError } from "@apollo/client/link/error";
import "firebase/auth";
import { Storage } from "@capacitor/storage";
import { trackEvent } from "utils/analytics/Segment";
import EVENTS from "utils/analytics/AnalyticEvents";
import { setMinAppVersion, currAppVersion, minAppVersionVar as currMinAppVersion } from "utils/update/update";
import { captureException } from "ErrorLogger";
import { GraphQLError } from "graphql";
import firebase from "../Firebase";

const httpLink = createHttpLink({
  uri: process.env.REACT_APP_HUMBL_DATA_URL,
  // credentials: 'same-origin',
});

const error = onError(({ operation, networkError, graphQLErrors, response }) => {
  trackEvent(EVENTS.API_ERROR, {
    operation,
    networkError,
    graphQLErrors,
  });
  if (networkError) {
    captureException(networkError);
  }

  if (graphQLErrors) {
    graphQLErrors.forEach((e) => {
      const sentryError = new Error();
      const name = e.extensions?.code ?? e.message;
      sentryError.name = name;
      sentryError.message = e.message;
      captureException(sentryError, generateGraphQLErrorMeta(e));
    });
  }

  if (response?.errors) {
    // eslint-disable-next-line no-param-reassign
    response.errors = response.errors.map((err) => ({
      ...err,
      humblErrorCode: err.extensions?.code,
    }));
  }
});

function generateGraphQLErrorMeta(graphqlError: GraphQLError) {
  return Object.entries(graphqlError).reduce((curr, next) => {
    const [k, v] = next;
    // Ignore message since we use that as the main error text anyways.
    if (k !== "message") {
      try {
        curr[k] = JSON.stringify(v);
      } catch (e) {
        captureException(e);
      }
    }

    return curr;
  }, {} as Record<string, string>);
}

const authLink = setContext(async (_, { headers }) => {
  // get the authentication token from firebase if it exists
  const token = await firebase.auth().currentUser?.getIdToken();
  // return the headers to the context so httpLink can read them
  Storage.migrate();
  const accountId = await Storage.get({ key: "defaultAccountId" });
  const headerObject = {
    ...headers,
    authorization: token ? `Bearer ${token}` : "",
    version: currAppVersion,
  };
  // If we have a default account in local storage to work with, send it along
  if (accountId.value) {
    headerObject.account = accountId.value;
  }
  return {
    headers: headerObject,
  };
});

const responseLink = new ApolloLink((operation, forward) => forward(operation).map((response) => {
    const context = operation.getContext();
    const minAppVersion: string = context?.response?.headers?.get("minimum-app-version") ?? "";

    if (minAppVersion !== currMinAppVersion()) {
      setMinAppVersion(minAppVersion);
    }

    return response;
  }));

const client = new ApolloClient({
  cache: new InMemoryCache(),
  uri: process.env.REACT_APP_HUMBL_DATA_URL,
  link: from([error, authLink, responseLink.concat(httpLink)]),
  defaultOptions: { query: { fetchPolicy: "cache-first" } },
});

export default client;
