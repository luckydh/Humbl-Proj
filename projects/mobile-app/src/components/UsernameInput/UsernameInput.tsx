import React, { FC, useCallback, useRef, useState } from "react";
import Input, { LabelContainer } from "components/Input/Input";
import { Message } from "components/Message/Message";
import { useTranslation } from "react-i18next";
import { loader } from "graphql.macro";

import { IsUsernameAvailableQuery } from "generated/graphql";
import useImperativeQuery from "hooks/useImperativeQuery";
import AwesomeDebouncePromise from "awesome-debounce-promise";
import Joi from "joi";

const USERNAME_AVAILABLE = loader("../../queries/USERNAMEAVAILABLE.gql");

export const USERNAME_MIN_LENGTH = 6;

export const UsernameSchema = Joi.string()
  .required()
  .min(6)
  .max(24)
  .pattern(/^[a-zA-Z\d]+([-][a-zA-Z\d]+)*$/);
interface UsernameInputProps {
  register: any;
  errors: any;
  trigger?: (payload?: string | string[]) => Promise<boolean>;
  onValidate?: (isValid: boolean) => void;
  ariaLabel?: string;
}

const UsernameInput: FC<UsernameInputProps> = ({ trigger, register, errors, onValidate, ariaLabel }) => {
  const { t } = useTranslation();
  const [loading, setLoading] = useState<boolean>(false);
  const [isAvailable, setIsAvailable] = useState<boolean | undefined>(undefined);

  const isUsernameAvailable = useImperativeQuery<IsUsernameAvailableQuery>(USERNAME_AVAILABLE);

  // wrapped in a ref to avoid recreating the debounced function everytime
  // https://github.com/slorber/awesome-debounce-promise#why-is-my-debouncing-function-always-firing-and-is-not-debounced
  const checkUsername = useRef(
    AwesomeDebouncePromise(async (username?: string) => {
      setLoading(true);

      const result = await isUsernameAvailable({
        keyword: username,
      });

      setLoading(false);
      setIsAvailable(result?.data?.isUsernameAvailable);

      if (onValidate) {
        onValidate(!!result?.data?.isUsernameAvailable);
      }
    }, 500)
  ).current;

  const onChange = useCallback(
    async (event: React.ChangeEvent<HTMLInputElement>) => {
      event.persist();
      if (trigger) {
        // triggers parent validation and waits for result
        const isValid = await trigger("userName");
        // only run if parent validation succeeds
        if (isValid) {
          checkUsername(event.target.value);
        } else {
          // this prevents the success message from previous
          // validations attempts from showing up when it shouldn't
          setIsAvailable(undefined);
        }
      }
    },
    [trigger, checkUsername]
  );

  const renderMessage = () => {
    if (loading) {
      return (
        <div className="text-white flex items-center text-sm rounded-md p-2 leading-5">
          {t("merchant-create-page.message.username-check")}
        </div>
      );
    }

    if (!errors.userName?.message) {
      if (isAvailable) {
        return (
          <div className="my-2 ">
            <Message variant="success">{t("user-name.success.this-username-is-available")}</Message>
          </div>
        );
      }

      if (typeof isAvailable !== "undefined" && !isAvailable) {
        return (
          <Message ariaLabel={ariaLabel && `${ariaLabel}ALERT_LABEL`} variant="error">
            {t("user-name.failure.this-username-is-not-available")}
          </Message>
        );
      }
    }

    return null;
  };

  return (
    <div className="mb-2">
      <LabelContainer ariaLabel={ariaLabel && `${ariaLabel}_LABEL`} label={t("merchant-create-page.label.user-name")} />
      <Input
        placeholder={t("merchant-create-page.placeholder.user-name")}
        name="userName"
        ariaLabel={ariaLabel}
        type="text"
        minLength={USERNAME_MIN_LENGTH}
        maxLength={24}
        onChange={onChange}
        register={register}
        errorMessage={errors.userName?.message}
      />
      {renderMessage()}
      <div
        aria-label={ariaLabel && `${ariaLabel}INSTRUCTIONS_LABEL`}
        className="pt-2 pb-4 u-1 text-white text-sm font-medium">
        {t("merchant-create-page.info.usernames-can-only-be-letters-numbers-hyphens")}
      </div>
    </div>
  );
};

export default UsernameInput;
