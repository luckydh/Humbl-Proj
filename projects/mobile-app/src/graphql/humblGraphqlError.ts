import { ApolloError } from "@apollo/client";
import { GraphQLError } from "graphql";
import i18n from "i18n";

export type HumblErrorCode = keyof typeof errorCodesTranslationKeys;

export interface HumblError {
  humblErrorCode: HumblErrorCode;
}

export interface HumblGraphQLError extends GraphQLError {
  humblErrorCode?: HumblErrorCode;
}

const errorCodesTranslationKeys = {
  SERVER_ERROR: "crypto-wallet.buy.order-preview.error-modal.fallback-message",
  IDENTITY_MISMATCH: "crypto-wallet.buy.order-preview.error-modal.identity-mismatch",
  PHONE_MISMATCH_INVALID: "crypto-wallet.buy.order-preview.error-modal.phone-mismatch-invalid",
  TOO_MUCH_ABUSE_SUSPICIOUS: "crypto-wallet.buy.order-preview.error-modal.abuse-suspicious",
  AUTHORIZATION_FAILED: "crypto-wallet.buy.order-preview.error-modal.auth-failed",
  ORDER_EXPIRED: "crypto-wallet.buy.order-preview.error-modal.order-expired",
  DECLINED_DO_NOT_HONOR: "crypto-wallet.buy.order-preview.error-modal.declined-do-not-honor",
  INSUFFICIENT_FUNDS: "crypto-wallet.buy.order-preview.error-modal.insufficient-funds",
  RESTRICTED: "crypto-wallet.buy.order-preview.error-modal.restricted",
  CARD_ISSUER_DECLINED: "crypto-wallet.buy.order-preview.error-modal.card-issuer-declined",
  LIMIT_REACHED: "crypto-wallet.buy.order-preview.error-modal.limit-reached",
  VALIDATION_ERROR: "crypto-wallet.buy.order-preview.error-modal.validation-error",
};

const fallback = errorCodesTranslationKeys.SERVER_ERROR;

export function getPotentialHumblError(graphQLErrors: ReadonlyArray<HumblGraphQLError>): HumblError | undefined {
  // This can sometimes trigger with 0 graphQLErrors hence this guard.
  const humblErrorCode = graphQLErrors[0]?.humblErrorCode;
  return humblErrorCode && { humblErrorCode };
}

export function getErrorMessage(error?: ApolloError) {
  if (!error) {
    return;
  }
  const humblError = getPotentialHumblError(error.graphQLErrors);
  return i18n.t(mapErrorCodeToTranslationKey(humblError?.humblErrorCode));
}

export function mapErrorCodeToTranslationKey(errorCode: HumblErrorCode = "SERVER_ERROR") {
  return errorCodesTranslationKeys[errorCode] ?? fallback;
}
