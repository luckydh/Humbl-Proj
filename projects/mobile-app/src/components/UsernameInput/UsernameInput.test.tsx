import React from "react";
import { render } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { MockedProvider } from "@apollo/client/testing";
import { IsUsernameAvailableDocument } from "generated/graphql";
import UsernameInput, { UsernameSchema } from "./UsernameInput";

jest.mock("react-i18next", () => ({
  // this mock makes sure any components using the translate hook can use it without a warning being shown
  useTranslation: () => ({
      t: (string: string) => string,
    }),
}));

const LABEL = "merchant-create-page.label.user-name";
const PLACEHOLDER = "merchant-create-page.placeholder.user-name";
const CONSTRAINTS = "merchant-create-page.info.usernames-can-only-be-letters-numbers-hyphens";
const LOADING = "merchant-create-page.message.username-check";
const AVAILABLE = "user-name.success.this-username-is-available";
const UNAVAILABLE = "user-name.failure.this-username-is-not-available";

describe("UsernameInput", () => {
  it("should render the default input", () => {
    const { queryByText, queryByPlaceholderText, asFragment } = render(
      <MockedProvider>
        <UsernameInput errors={{}} register={() => null} />
      </MockedProvider>
    );

    expect(queryByText(LABEL)).toBeInTheDocument();
    expect(queryByPlaceholderText(PLACEHOLDER)).toBeInTheDocument();
    expect(queryByText(CONSTRAINTS)).toBeInTheDocument();

    expect(queryByText(LOADING)).not.toBeInTheDocument();
    expect(queryByText(AVAILABLE)).not.toBeInTheDocument();
    expect(queryByText(UNAVAILABLE)).not.toBeInTheDocument();

    expect(asFragment()).toMatchSnapshot();
  });

  it("should display the success message if username is available", async () => {
    const usernameMock = {
      request: {
        query: IsUsernameAvailableDocument,
        variables: { keyword: "test" },
      },
      result: {
        data: { isUsernameAvailable: true },
      },
    };

    const onValidate = jest.fn();

    const { findByText, getByPlaceholderText, asFragment } = render(
      <MockedProvider mocks={[usernameMock]}>
        <UsernameInput errors={{}} register={() => null} trigger={async () => true} onValidate={onValidate} />
      </MockedProvider>
    );

    userEvent.type(getByPlaceholderText(PLACEHOLDER), "test");

    expect(await findByText(AVAILABLE)).toBeInTheDocument();
    expect(onValidate).toBeCalledWith(true);
    expect(asFragment()).toMatchSnapshot();
  });

  it("should display the error message if the username is not available", async () => {
    const usernameMock = {
      request: {
        query: IsUsernameAvailableDocument,
        variables: { keyword: "test" },
      },
      result: {
        data: { isUsernameAvailable: false },
      },
    };

    const onValidate = jest.fn();

    const { findByText, getByPlaceholderText, asFragment } = render(
      <MockedProvider mocks={[usernameMock]}>
        <UsernameInput errors={{}} register={() => null} trigger={async () => true} onValidate={onValidate} />
      </MockedProvider>
    );

    userEvent.type(getByPlaceholderText(PLACEHOLDER), "test");

    expect(await findByText(UNAVAILABLE)).toBeInTheDocument();
    expect(onValidate).toBeCalledWith(false);
    expect(asFragment()).toMatchSnapshot();
  });

  it("should display the error message if validation fails", () => {
    const errors = {
      userName: {
        message: "validation error",
      },
    };

    const { queryByText, asFragment } = render(
      <MockedProvider>
        <UsernameInput errors={errors} register={() => null} />
      </MockedProvider>
    );

    expect(queryByText("validation error")).toBeInTheDocument();
    expect(asFragment()).toMatchSnapshot();
  });

  it("should not display other messages if it has validation errors", () => {
    const errors = {
      userName: {
        message: "validation error",
      },
    };

    const usernameMock = {
      request: {
        query: IsUsernameAvailableDocument,
        variables: { keyword: "test" },
      },
      result: {
        data: { isUsernameAvailable: true },
      },
    };

    const { queryByText, getByPlaceholderText } = render(
      <MockedProvider mocks={[usernameMock]}>
        <UsernameInput errors={errors} register={() => null} trigger={async () => true} />
      </MockedProvider>
    );

    userEvent.type(getByPlaceholderText(PLACEHOLDER), "test");

    expect(queryByText("validation error")).toBeInTheDocument();

    expect(queryByText(LOADING)).not.toBeInTheDocument();
    expect(queryByText(AVAILABLE)).not.toBeInTheDocument();
    expect(queryByText(UNAVAILABLE)).not.toBeInTheDocument();
  });
});

describe("UsernameSchema", () => {
  test.each([
    // [input, expected result (true for valid, false for invalid)]
    ["username", true],
    ["username123", true],
    ["user_name", false],
    ["user-name", true],
    ["user.name", false],
    ["user_name.123", false],
    ["user_name-123", false],
    ["user-name.123", false],
    ["user-name_123", false],
    ["user.name-123", false],
    ["user.name_123", false],
    ["u.s.e.r.n.a.m.e", false],
    ["u-s-e-r-n-a-m-e", true],
    ["u_s_e_r_n_a_m_e", false],
    ["user__name", false],
    ["user--name", false],
    ["user..name", false],
    ["user-._name", false],
    [".username", false],
    ["-username", false],
    ["_username", false],
    ["username.", false],
    ["username-", false],
    ["username_", false],
    ["", false],
    ["u", false],
    ["us", false],
    ["with-24-char-should-pass", true],
    ["usernames-with-more-than-24-characters-shouldnt-be-allowed", false],
  ])('should validate "%s" to %s', (username, expected) => {
    const result = UsernameSchema.validate(username);
    expect(!result?.error).toEqual(expected);
  });
});
