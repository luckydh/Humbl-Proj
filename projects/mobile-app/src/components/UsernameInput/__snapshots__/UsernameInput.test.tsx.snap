// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UsernameInput should display the error message if the username is not available 1`] = `
<DocumentFragment>
  <div
    class="mb-2"
  >
    <div
      class="ml-2"
    >
      <label
        class="block text-xs uppercase text-white opacity-80 font-medium mb-1"
      >
        merchant-create-page.label.user-name
      </label>
    </div>
    <div
      class="mb-2"
    >
      <div
        class="py-0 w-full flex items-center rounded-lg select-none bg-blue border-white border-2 border-solid outline-none"
      >
        <input
          class="text-white placeholder-white-faded w-full rounded-lg bg-transparent text-md sm:text-xl outline-none border-none focus:border-transparent focus:outline-none focus:ring-0 focus:placeholder-transparent"
          maxlength="24"
          minlength="6"
          name="userName"
          placeholder="merchant-create-page.placeholder.user-name"
          style="height: 48px; font-size: 20px;"
          type="text"
          value=""
        />
      </div>
    </div>
    <div
      class="flex items-center text-sm rounded-md p-2 leading-5 text-gray-900 border-red bg-red-light"
      style="color: rgb(59, 91, 123);"
    >
      user-name.failure.this-username-is-not-available
      <div
        class="pl-2 ml-auto"
      />
    </div>
    <div
      class="pt-2 pb-4 u-1 text-white text-sm font-medium"
    >
      merchant-create-page.info.usernames-can-only-be-letters-numbers-hyphens
    </div>
  </div>
</DocumentFragment>
`;

exports[`UsernameInput should display the error message if validation fails 1`] = `
<DocumentFragment>
  <div
    class="mb-2"
  >
    <div
      class="ml-2"
    >
      <label
        class="block text-xs uppercase text-white opacity-80 font-medium mb-1"
      >
        merchant-create-page.label.user-name
      </label>
    </div>
    <div
      class="mb-2"
    >
      <div
        class="py-0 w-full flex items-center rounded-lg select-none bg-blue border-white border-2 border-solid outline-none mb-2"
      >
        <input
          class="text-white placeholder-white-faded w-full rounded-lg bg-transparent text-md sm:text-xl outline-none border-none focus:border-transparent focus:outline-none focus:ring-0 focus:placeholder-transparent"
          maxlength="24"
          minlength="6"
          name="userName"
          placeholder="merchant-create-page.placeholder.user-name"
          style="height: 48px; font-size: 20px;"
          type="text"
          value=""
        />
      </div>
      <div>
        <div
          class="flex items-center text-sm rounded-md p-2 leading-5 text-gray-900 border-red bg-red-light"
          style="color: rgb(59, 91, 123);"
        >
          validation error
          <div
            class="pl-2 ml-auto"
          />
        </div>
      </div>
    </div>
    <div
      class="pt-2 pb-4 u-1 text-white text-sm font-medium"
    >
      merchant-create-page.info.usernames-can-only-be-letters-numbers-hyphens
    </div>
  </div>
</DocumentFragment>
`;

exports[`UsernameInput should display the success message if username is available 1`] = `
<DocumentFragment>
  <div
    class="mb-2"
  >
    <div
      class="ml-2"
    >
      <label
        class="block text-xs uppercase text-white opacity-80 font-medium mb-1"
      >
        merchant-create-page.label.user-name
      </label>
    </div>
    <div
      class="mb-2"
    >
      <div
        class="py-0 w-full flex items-center rounded-lg select-none bg-blue border-white border-2 border-solid outline-none"
      >
        <input
          class="text-white placeholder-white-faded w-full rounded-lg bg-transparent text-md sm:text-xl outline-none border-none focus:border-transparent focus:outline-none focus:ring-0 focus:placeholder-transparent"
          maxlength="24"
          minlength="6"
          name="userName"
          placeholder="merchant-create-page.placeholder.user-name"
          style="height: 48px; font-size: 20px;"
          type="text"
          value=""
        />
      </div>
    </div>
    <div
      class="my-2 "
    >
      <div
        class="flex items-center text-sm rounded-md p-2 leading-5 bg-green-100 text-green-800 border-green-800"
      >
        user-name.success.this-username-is-available
        <div
          class="pl-2 ml-auto"
        >
          <svg
            height="21.04"
            width="28.054"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              class="fill-current"
              d="M0 11.415l9.607 9.625L28.054 2.612 25.406 0l-15.8 15.78-7-7z"
            />
          </svg>
        </div>
      </div>
    </div>
    <div
      class="pt-2 pb-4 u-1 text-white text-sm font-medium"
    >
      merchant-create-page.info.usernames-can-only-be-letters-numbers-hyphens
    </div>
  </div>
</DocumentFragment>
`;

exports[`UsernameInput should render the default input 1`] = `
<DocumentFragment>
  <div
    class="mb-2"
  >
    <div
      class="ml-2"
    >
      <label
        class="block text-xs uppercase text-white opacity-80 font-medium mb-1"
      >
        merchant-create-page.label.user-name
      </label>
    </div>
    <div
      class="mb-2"
    >
      <div
        class="py-0 w-full flex items-center rounded-lg select-none bg-blue border-white border-2 border-solid outline-none"
      >
        <input
          class="text-white placeholder-white-faded w-full rounded-lg bg-transparent text-md sm:text-xl outline-none border-none focus:border-transparent focus:outline-none focus:ring-0 focus:placeholder-transparent"
          maxlength="24"
          minlength="6"
          name="userName"
          placeholder="merchant-create-page.placeholder.user-name"
          style="height: 48px; font-size: 20px;"
          type="text"
          value=""
        />
      </div>
    </div>
    <div
      class="pt-2 pb-4 u-1 text-white text-sm font-medium"
    >
      merchant-create-page.info.usernames-can-only-be-letters-numbers-hyphens
    </div>
  </div>
</DocumentFragment>
`;
